File: packages/antgrid-wire/tests/relay-envelope-vectors.test.ts
// Cross-language pin for the relay control envelopes: every entry in
// evals/fixtures/relay-envelope-vectors.json must parse through the Zod
// unions, and every union variant / ErrorCode must have an entry. The Dart
// side (packages/antgrid_relay_client/test/relay_envelope_vectors_test.dart)
// runs the same fixture through the hand-mirrored schemas. Adding a wire
// variant without regenerating the fixture fails HERE; regenerating without
// updating the Dart mirror fails THERE.
import { test, expect } from "bun:test";
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { ClientMessage, ErrorCode, ServerMessage } from "../src/relay-protocol";

type Vector = { name: string; json: Record<string, unknown> };

const fixture = JSON.parse(
  readFileSync(resolve(import.meta.dir, "../../../evals/fixtures/relay-envelope-vectors.json"), "utf8"),
) as { server: Vector[]; client: Vector[] };

test("every server vector parses through ServerMessage", () => {
  for (const v of fixture.server) {
    expect(() => ServerMessage.parse(v.json), v.name).not.toThrow();
  }
});

test("every client vector parses through ClientMessage", () => {
  for (const v of fixture.client) {
    expect(() => ClientMessage.parse(v.json), v.name).not.toThrow();
  }
});

test("every ServerMessage variant has a vector", () => {
  const variants = ServerMessage.options.map((o) => o.shape.type.value as string);
  const covered = new Set(fixture.server.map((v) => v.json.type));
  expect([...covered].sort()).toEqual([...new Set(variants)].sort());
});

test("every ClientMessage variant has a vector", () => {
  const variants = ClientMessage.options.map((o) => o.shape.type.value as string);
  const covered = new Set(fixture.client.map((v) => v.json.type));
  expect([...covered].sort()).toEqual([...new Set(variants)].sort());
});

test("every ErrorCode has an error vector", () => {
  const covered = fixture.server
    .filter((v) => v.json.type === "error")
    .map((v) => v.json.code as string);
  expect(covered.sort()).toEqual([...ErrorCode.options].sort());
});

File: packages/antgrid_relay_client/test/relay_envelope_vectors_test.dart
// Cross-language pin for the relay control envelopes: this package mirrors
// the antgrid-wire Zod schemas BY HAND, and this test runs the shared golden
// fixture through that mirror. The TS side
// (packages/antgrid-wire/tests/relay-envelope-vectors.test.ts) guarantees the
// fixture covers every union variant and ErrorCode, so a wire-schema change
// that isn't mirrored here fails at test time instead of at runtime on a
// device. Regenerate the fixture:
//   cd packages/antgrid-wire && bun run scripts/gen-envelope-vectors.ts
import 'dart:convert';
import 'dart:io';

import 'package:antgrid_relay_client/antgrid_relay_client.dart';
import 'package:test/test.dart';

void main() {
  final v =
      jsonDecode(
            File(
              '../../evals/fixtures/relay-envelope-vectors.json',
            ).readAsStringSync(),
          )
          as Map<String, dynamic>;

  final server = (v['server'] as List).cast<Map<String, dynamic>>();
  final client = (v['client'] as List).cast<Map<String, dynamic>>();

  group('server vectors (relay → client)', () {
    for (final vec in server) {
      final name = vec['name'] as String;
      final dart = vec['dart'] as String;
      final json = (vec['json'] as Map<String, dynamic>);

      test('$name → $dart', () {
        final parsed = parseRelayMessage(json);
        if (dart == 'tolerated') {
          // Agent-only frames: the tolerance contract says fall through to
          // null, never throw — an app must ignore what it doesn't speak.
          expect(parsed, isNull);
          return;
        }
        expect(parsed, isNotNull, reason: 'mirror failed to parse $name');
        switch (json['type'] as String) {
          case 'welcome':
            final m = parsed as WelcomeMessage?;
            expect(m!.deviceId, json['deviceId']);
            expect(m.epoch, json['epoch']);
            expect(m.serverTime, json['serverTime']);
          case 'stream-opened':
            expect(
              (parsed as StreamOpenedMessage?)!.streamId,
              json['streamId'],
            );
          case 'stream-closed':
            expect(
              (parsed as StreamClosedMessage?)!.streamId,
              json['streamId'],
            );
          case 'error':
            final m = parsed as ErrorMessage?;
            expect(m!.code, json['code']);
            expect(m.message, json['message']);
            expect(m.retryable, json['retryable']);
            expect(m.ref, json['ref']);
            expect(m.serverTime, json['serverTime']);
          case 'peer-online':
            expect((parsed as PeerOnlineMessage?)!.peerId, json['peerId']);
          case 'peer-offline':
            expect((parsed as PeerOfflineMessage?)!.peerId, json['peerId']);
          default:
            fail(
              'vector $name marked parsed but has no field assertions — '
              'add a case when mirroring a new server frame',
            );
        }
      });
    }
  });

  group('client vectors (client → relay)', () {
    for (final vec in client.where((c) => c['dartEmits'] == true)) {
      final name = vec['name'] as String;
      final json = (vec['json'] as Map<String, dynamic>);

      test('$name emitted byte-identical', () {
        final Map<String, dynamic> emitted = switch (json['type'] as String) {
          'hello' => HelloMessage(
            deviceType: json['deviceType'] as String,
            deviceId: json['deviceId'] as String,
            name: json['name'] as String,
            publicKey: json['publicKey'] as String,
            epoch: json['epoch'] as int,
            licenseToken: json['licenseToken'] as String,
            ts: json['ts'] as String,
            nonce: json['nonce'] as String,
            sig: json['sig'] as String,
          ).toJson(),
          'stream-open' => StreamOpenMessage(
            streamId: json['streamId'] as String,
          ).toJson(),
          'stream-close' => StreamCloseMessage(
            streamId: json['streamId'] as String,
          ).toJson(),
          _ => fail(
            'vector $name marked dartEmits but has no constructor '
            'case — add one when the Dart client learns to send it',
          ),
        };
        expect(emitted, equals(json));
      });
    }
  });
}
